prov/verbs/src/verbs_mr.c

```diff
@@ -130,37 +154,292 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
+int fi_ibv_mr_regattr_check_args(struct fid *fid,
+				 const struct fi_mr_attr *attr,
+				 uint64_t flags)
 {
-	if (count > VERBS_MR_IOV_LIMIT) {
+	if (OFI_UNLIKELY(flags))
+		return -FI_EBADFLAGS;
+	if (OFI_UNLIKELY(fid->fclass != FI_CLASS_DOMAIN))
+		return -FI_EINVAL;
+	if (OFI_UNLIKELY(attr->iov_count > VERBS_MR_IOV_LIMIT)) {
 		VERBS_WARN(FI_LOG_FABRIC,
 			   "iov count > %d not supported\n",
 			   VERBS_MR_IOV_LIMIT);
 		return -FI_EINVAL;
 	}
-	return fi_ibv_mr_reg(fid, iov->iov_base, iov->iov_len, access, offset,
-			requested_key, flags, mr, context);
+	return FI_SUCCESS;
+}
+
+static inline int
+fi_ibv_mr_ofi2ibv_access(uint64_t ofi_access, struct fi_ibv_domain *domain)
+{
+	int ibv_access = 0;
+
+	/* Enable local write access by default for FI_EP_RDM which hides local
+	 * registration requirements. This allows to avoid buffering or double
+	 * registration */
+	if (!(domain->info->caps & FI_LOCAL_MR) &&
+	    !(domain->info->domain_attr->mr_mode & FI_MR_LOCAL))
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE;
+
+	/* Local read access to an MR is enabled by default in verbs */
+	if (ofi_access & FI_RECV)
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE;
+
+	/* iWARP spec requires Remote Write access for an MR that is used
+	 * as a data sink for a Remote Read */
+	if (ofi_access & FI_READ) {
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE;
+		if (domain->verbs->device->transport_type == IBV_TRANSPORT_IWARP)
+			ibv_access |= IBV_ACCESS_REMOTE_WRITE;
+	}
+
+	if (ofi_access & FI_WRITE)
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE;
+
+	if (ofi_access & FI_REMOTE_READ)
+		ibv_access |= IBV_ACCESS_REMOTE_READ;
+
+	/* Verbs requires Local Write access too for Remote Write access */
+	if (ofi_access & FI_REMOTE_WRITE)
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE |
+			      IBV_ACCESS_REMOTE_WRITE |
+			      IBV_ACCESS_REMOTE_ATOMIC;
+
+	return ibv_access;
+}
+
+static inline struct fi_ibv_mem_desc *
+fi_ibv_mr_common_cache_reg(struct fi_ibv_domain *domain,
+			 struct fi_mr_attr *attr)
+{
+	struct fi_ibv_mem_desc *md;
+	struct ofi_mr_entry *entry;
+	int ret;
+
+	ret = ofi_mr_cache_search(&domain->cache, attr, &entry);
+	if (OFI_UNLIKELY(ret))
+		return NULL;
+
+	md = (struct fi_ibv_mem_desc *)entry->data;
+	md->entry = entry;
+
+	return md;
+}
+
+static inline
+void fi_ibv_common_cache_dereg(struct fi_ibv_mem_desc *md)
+{
+	ofi_mr_cache_delete(&md->domain->cache, md->entry);
+}
+
+static inline
+int fi_ibv_mr_internal_reg(struct fi_ibv_domain *domain, void *buf,
+			   size_t len, uint64_t access,
+			   struct fi_ibv_mem_desc *md)
+{
+	md->domain = domain;
+	md->len = len;
+	md->mr = fi_ibv_mr_reg_ibv_mr(domain, buf, len,
+				      fi_ibv_mr_ofi2ibv_access(access,
+							       domain));
+	if (OFI_UNLIKELY(!md->mr))
+		return -errno;
+	return FI_SUCCESS;
+}
+
+int fi_ibv_mr_direct_reg(struct fi_ibv_domain *domain, void *buf,
+			 size_t len, uint64_t access,
+			 struct fi_ibv_mem_desc *md)
+{
+	return fi_ibv_mr_internal_reg(domain, buf, len, access, md);
+}
```

Comment:
Just curious why this extra function is needed. It does nothing but call fi_ibv_mr_internal_reg.